src/dag.c

```diff
@@ -168,6 +168,107 @@ void RedisAI_DagRunSession_ModelRun_Step(RedisAI_RunInfo *rinfo, RAI_DagOp *curr
   return;
 }
 
+/**
+ * Execution of a MODELRUN DAG step.
+ * If an error occurs, it is recorded in the DagOp struct.
+ *
+ * @param rinfo context in which RedisAI blocking commands operate.
+ * @param currentOp MODELRUN DagOp to be executed
+ * @return
+ */
+void RedisAI_BatchedDagRunSession_ModelRun_Step(RedisAI_RunInfo **batched_rinfo, RAI_DagOp **currentOps) {
+
+  int n_rinfo = array_len(batched_rinfo);
+
+  RAI_ModelRunCtx *mctxs[n_rinfo];
+
+  for (int i=0; i<n_rinfo; i++) {
+    RedisAI_RunInfo *rinfo = batched_rinfo[i];
+    RAI_DagOp *currentOp = currentOps[i];
+
+    pthread_mutex_lock(rinfo->dagMutex);
+
+    uint n_inkeys = array_len(currentOp->inkeys);
+    uint n_outkeys = array_len(currentOp->outkeys);
+
+    RAI_Tensor* inputTensors[n_inkeys];
+    for (uint i=0; i<n_inkeys; i++) {
+      RAI_Tensor *inputTensor;
+      const int get_result = RAI_getTensorFromLocalContext(
+          NULL, rinfo->dagTensorsContext, RedisModule_StringPtrLen(currentOp->inkeys[i], NULL), &inputTensor, currentOp->err);
+      if (get_result == REDISMODULE_ERR) {
+        // We check for this outside the function
+        // this check cannot be covered by tests
+        currentOp->result = REDISMODULE_ERR;
+        pthread_mutex_unlock(rinfo->dagMutex);
+        return;
+      }
+      inputTensors[i] = inputTensor;
+    }
+
+    for (uint i=0; i<n_inkeys; i++) {
+      const char *opname = NULL;
+      if (currentOp->mctx->model->inputs) {
+        opname = currentOp->mctx->model->inputs[i];
+      }
+      RAI_ModelRunCtxAddInput(currentOp->mctx, opname, inputTensors[i]);
+    }
+
+    for (uint i=0; i<n_outkeys; i++) {
+      const char *opname = NULL;
+      if (currentOp->mctx->model->inputs) {
+        opname = currentOp->mctx->model->outputs[i];
+      }
+      RAI_ModelRunCtxAddOutput(currentOp->mctx, opname);
+    }
+
+    mctxs[i] = currentOp->mctx;
+
+    pthread_mutex_unlock(rinfo->dagMutex);
+  }
+
+  RAI_Error err = {0};
+  const long long start = ustime();
+  int result = RAI_ModelRun(mctxs, n_rinfo, &err);
+  const long long end = ustime();
+
+  long long duration = end - start;
+
+  for (int i=0; i<n_rinfo; i++) {
+    RedisAI_RunInfo *rinfo = batched_rinfo[i];
+    RAI_DagOp *currentOp = currentOps[i];
+
+    pthread_mutex_lock(rinfo->dagMutex);
+
+    if (result == REDISMODULE_ERR) {
+      currentOp->result = result;
+      RAI_SetError(currentOp->err, err.code, err.detail);
+      pthread_mutex_unlock(rinfo->dagMutex);
```

Comment:
Is there a possibility to move the locking after the error set? Should it be locking the Dag?

Reply:
We can remove the lock if we turn `result` into an `atomic` (which we should do), since `result` is checked by others.

Reply:
Done